common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java

```diff
@@ -84,57 +103,162 @@ private CoverageSearchEvent getLastEvent(CoveragePeriod period) {
     //      of the current search. This needs to run after the completion of every search
     @Override
-    public void deletePreviousSearch(int searchId) {
-        CoveragePeriod period = findCoveragePeriod(searchId);
-        Optional<CoverageSearchEvent> searchEvent = coverageSearchEventRepo.findSearchWithOffset(searchId, 1);
+    public List<CoverageSummary> pageCoverage(int pageNumber, int pageSize, List<Integer> coveragePeriodIds) {
 
-        searchEvent.ifPresent(coverageSearchEvent ->
-                coverageRepo.removeAllByCoveragePeriodAndCoverageSearchEvent(period, coverageSearchEvent));
+        CoveragePeriod coveragePeriod = coveragePeriodRepo.getOne(coveragePeriodIds.get(0));
+
+        Map<String, List<Object[]>> coverageMembership = findCoverageMemberships(pageNumber, pageSize, coveragePeriodIds);
+
+        return coverageMembership.entrySet().stream()
+                .map(obj -> summarizeCoverageMembership(coveragePeriod.getContract(), obj))
+                .collect(toList());
+    }
+
+    @Override
+    public List<CoverageSummary> pageCoverage(int pageNumber, int pageSize, Integer... coveragePeriodIds) {
+        return pageCoverage(pageNumber, pageSize, List.of(coveragePeriodIds));
+    }
+
+    /**
+     * Find page of beneficiaries then look up all of their coverage information for a range of identified coverage
+     * periods.
+     */
+    private Map<String, List<Object[]>> findCoverageMemberships(int pageNumber, int pageSize, List<Integer> coveragePeriodIds) {
+        List<CoveragePeriod> coveragePeriods = coveragePeriodRepo.findAllById(coveragePeriodIds);
+
+        // Look up a page of beneficiaries sorted by id
+        Page<String> beneficiaryPage = coverageRepo.findActiveBeneficiaryIds(coveragePeriods, PageRequest.of(pageNumber, pageSize));
+        List<String> beneficiaries = beneficiaryPage.getContent();
+
+        // Look up coverage information for BATCH_SELECT_SIZE beneficiaries at a time
+        // and populate into hash map
+        Map<String, List<Object[]>> coverageMemberships = new HashMap<>(beneficiaries.size());
+        int startIndex = 0;
+        while (startIndex < beneficiaries.size()) {
+            int endIndex = Math.min(startIndex + BATCH_SELECT_SIZE, beneficiaries.size());
+            List<Object[]> rawResults = coverageRepo.findCoverageInformation(coveragePeriodIds, beneficiaries.subList(startIndex, endIndex));
+            Map<String, List<Object[]>> coverages = rawResults.stream()
+                    .collect(groupingBy(raw -> (String) raw[0]));
+            coverageMemberships.putAll(coverages);
+
+            startIndex = endIndex;
+        }
+
+        return coverageMemberships;
+    }
+
+    /**
+     * Summarize the coverage of one beneficiary for
+     */
+    private CoverageSummary summarizeCoverageMembership(Contract contract, Map.Entry<String, List<Object[]>> membershipInfo) {
```

Comment:
This method seems to have quite a bit of complexity that I'm not convinced is needed.

Is the list of **membershipInfo** sorted?  Would it simplify processing if it was?  Since the list refers to months, maximum number is 12, correct?

If so, then performance over processing 12 items can't be a concern and clarity should be the goal.

Reply:
Both months and years, you can wrap years.